### PR TITLE
Convert to markdown: jit.rst

### PR DESCRIPTION
**Pull request by thatgeeman**
Fixes https://github.com/pytorch/pytorch/issues/155024

Related comment https://github.com/pytorch/pytorch/issues/155404#issuecomment-2960752064

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @xmfan @svekars @sekyondaMeta @AlannaBurke

@pytorchbot label "topic: docs"
@pytorchbot label "topic: not user facing"

### CODE REVIEW

File: docs/source/jit.md [hunk 1]
```diff
@@ -0,0 +1,901 @@
...
+See {ref}`builtin-functions` for a full reference of supported functions.
+
+### PyTorch Functions and Modules
+
+TorchScript supports a subset of the tensor and neural network
+functions that PyTorch provides. Most methods on Tensor as well as functions in
+the `torch` namespace, all functions in `torch.nn.functional` and
+most modules from `torch.nn` are supported in TorchScript.
+
+See {ref}`jit-unsupported` for a list of unsupported PyTorch functions and modules.
```
Comment (svekars) [inline docs/source/jit.md]:
```suggestion
See {ref}`jit_unsupported` for a list of unsupported PyTorch functions and modules.
```

File: docs/source/jit.md [hunk 2]
```diff
@@ -0,0 +1,901 @@
...
+    jit_language_reference
+
+
+.. toctree::
+    :maxdepth: 1
+
+    jit_language_reference_v2
+
+
+.. contents:: :local:
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion
```{contents}
:local:
:depth: 2
```
```

File: docs/source/jit.md [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+```{eval-rst}
+.. toctree::
+   :maxdepth: 1
+   :caption: Builtin Functions
+   :hidden:
+
+   torch.jit.supported_ops <jit_builtin_functions>
+
+
+.. toctree::
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion
```{toctree}
```

File: docs/source/jit.md [hunk 4]
```diff
@@ -0,0 +1,901 @@
...
+    :maxdepth: 1
+    :caption: Language Reference
+    :hidden:
+
+    jit_language_reference
+
+
+.. toctree::
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion
```{toctree}
```

wrap these in {toctree} directives

File: docs/source/jit.md [hunk 5]
```diff
@@ -0,0 +1,901 @@
...
+    :maxdepth: 1
+
+    jit_language_reference_v2
+
+
+.. contents:: :local:
+    :depth: 2
+
+.. automodule:: torch.jit
```
Comment (nirajkamal) [inline docs/source/jit.md]: This can be wrapped in its own {eval-rst} directive

File: docs/source/jit.md [hunk 6]
```diff
@@ -0,0 +1,901 @@
...
+of a part of a model.
+
+Scripted functions can call traced functions. This is particularly useful when you need
+to use control-flow around a simple feed-forward model. For instance the beam search
+of a sequence to sequence model will typically be written in script but can call an
+encoder module generated using tracing.
+
+```{eval-rst}
+.. testsetup::
+
```
Comment (nirajkamal) [inline docs/source/jit.md]: recommend to be wrapped in code block ```python code block wrapper instead of eval-rst

File: docs/source/jit.md [hunk 7]
```diff
@@ -0,0 +1,901 @@
...
+        obj.__module__ = 'FakeMod'
+        return original_trace(obj, *args, **kwargs)
+
+    torch.jit.trace = trace_wrapper
+```
+
+Example (calling a traced function in script):
+
+```{eval-rst}
+.. testcode::
```
Comment (nirajkamal) [inline docs/source/jit.md]: recommend to be wrapped in ```python code block wrapper instead of eval-rst

File: docs/source/jit.md [hunk 8]
```diff
@@ -0,0 +1,901 @@
...
+```
+
+Traced functions can call script functions. This is useful when a small part of
+a model requires some control-flow even though most of the model is just a feed-forward
+network. Control-flow inside of a script function called by a traced function is
+preserved correctly.
+
+Example (calling a script function in a traced function):
+
+```{eval-rst}
```
Comment (nirajkamal) [inline docs/source/jit.md]: recommend to be inside a python code block wrapper instead

File: docs/source/jit.md [hunk 9]
```diff
@@ -0,0 +1,901 @@
...
+
+    traced_bar = torch.jit.trace(bar, (torch.rand(3), torch.rand(3), torch.rand(3)))
+```
+
+This composition also works for `nn.Module`s as well, where it can be used to generate
+a submodule using tracing that can be called from the methods of a script module.
+
+Example (using a traced module):
+
+```{eval-rst}
```
Comment (nirajkamal) [inline docs/source/jit.md]: recommend to be inside a python code block wrapper instead

File: docs/source/jit.md [hunk 10]
```diff
@@ -0,0 +1,901 @@
...
+we first traced it and when we traced it with the `check_inputs`. Indeed,
+the loop within the body of `loop_in_traced_fn` depends on the shape
+of the input `x`, and thus when we try another `x` with a different
+shape, the trace differs.
+
+In this case, data-dependent control flow like this can be captured using
+{func}`torch.jit.script` instead:
+
+```{eval-rst}
+.. testcode::
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a code block ```python block instead

File: docs/source/jit.md [hunk 11]
```diff
@@ -0,0 +1,901 @@
...
+}
+```
+
+#### Tracer Warnings
+
+The tracer produces warnings for several problematic patterns in traced
+computation. As an example, take a trace of a function that contains an
+in-place assignment on a slice (a view) of a Tensor:
+
+```{eval-rst}
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a code block instead

File: docs/source/jit.md [hunk 12]
```diff
@@ -0,0 +1,901 @@
...
+graph(%0 : Float(3, 4)) {
+    return (%0);
+}
+```
+
+We can fix this by modifying the code to not use the in-place update, but
+rather build up the result tensor out-of-place with `torch.cat`:
+
+```{eval-rst}
+.. testcode::
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a code block

File: docs/source/jit.md [hunk 13]
```diff
@@ -0,0 +1,901 @@
...
+- The `@torch.jit.script_method` decorator
+- Classes that inherit from `torch.jit.ScriptModule`
+- The `torch.jit.Attribute` wrapper class
+- The `__constants__` array
+- The `torch.jit.annotate` function
+
+#### Modules
+
+```{eval-rst}
+.. warning::
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a ```{warning} block

File: docs/source/jit.md [hunk 14]
```diff
@@ -0,0 +1,901 @@
...
+2. `torch.jit.script(nn_module_instance)` is now the preferred way to create
+   {class}`ScriptModule`, instead of inheriting from `torch.jit.ScriptModule`.
+   These changes combine to provide a simpler, easier-to-use API for converting
+   your `nn.Module`s into {class}`ScriptModule`, ready to be optimized and executed in a
+   non-Python environment.
+
+The new usage looks like this:
+
+```{eval-rst}
+.. testcode::
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a code block instead

File: docs/source/jit.md [hunk 15]
```diff
@@ -0,0 +1,901 @@
...
+specific device, so casting an already-loaded model may have unexpected
+effects. Casting the model _before_ saving it ensures that the tracer has
+the correct device information.
+
+Q: How do I store attributes on a {class}`ScriptModule`?
+
+Say we have a model like:
+
+```{eval-rst}
+.. testcode::
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a code block instead

File: docs/source/jit.md [hunk 16]
```diff
@@ -0,0 +1,901 @@
...
+.. autofunction:: export
+```
+
+#### Functions
+
+Functions don't change much, they can be decorated with {func}`@torch.jit.ignore <torch.jit.ignore>` or {func}`@torch.jit.unused <torch.jit.unused>` if needed.
+
+```{eval-rst}
+.. testcode::
+    # Same behavior as pre-PyTorch 1.2
```
Comment (nirajkamal) [inline docs/source/jit.md]: code block instead of eval-rst

File: docs/source/jit.md [hunk 17]
```diff
@@ -0,0 +1,901 @@
...
+    # Doesn't do anything, this function is already
+    # the main entry point
+    @torch.jit.export
+    def some_fn4():
+        return 2
+```
+
+#### TorchScript Classes
+
+```{eval-rst}
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion
```{warning}
```

File: docs/source/jit.md [hunk 18]
```diff
@@ -0,0 +1,901 @@
...
+.. warning::
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion
```

File: docs/source/jit.md [hunk 19]
```diff
@@ -0,0 +1,901 @@
...
+The TorchScript compiler needs to know the types of `module attributes`. Most types
+can be inferred from the value of the member. Empty lists and dicts cannot have their
+types inferred and must have their types annotated with [PEP 526-style](https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations) class annotations.
+If a type cannot be inferred and is not explicitly annotated, it will not be added as an attribute
+to the resulting {class}`ScriptModule`
+
+Old API:
+
+```{eval-rst}
+.. testcode::
```
Comment (nirajkamal) [inline docs/source/jit.md]: can be wrapped in a code block instead

File: docs/source/jit.md [hunk 20]
```diff
@@ -0,0 +1,901 @@
...
+```
+
+### Fusion Backends
+
+There are a couple of fusion backends available to optimize TorchScript execution. The default fuser on CPUs is NNC, which can perform fusions for both CPUs and GPUs. The default fuser on GPUs is NVFuser, which supports a wider range of operators and has demonstrated generated kernels with improved throughput. See the [NVFuser documentation](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/codegen/cuda/README.md) for more details on usage and debugging.
+
+### References
+
+```{eval-rst}
+.. toctree::
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion
```{toctree}
```

File: docs/source/jit.md [hunk 21]
```diff
@@ -0,0 +1,901 @@
...
+        return x, b
+```
+
+### Fusion Backends
+
+There are a couple of fusion backends available to optimize TorchScript execution. The default fuser on CPUs is NNC, which can perform fusions for both CPUs and GPUs. The default fuser on GPUs is NVFuser, which supports a wider range of operators and has demonstrated generated kernels with improved throughput. See the [NVFuser documentation](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/codegen/cuda/README.md) for more details on usage and debugging.
+
+### References
+
+```{eval-rst}
```
Comment (nirajkamal) [inline docs/source/jit.md]:
```suggestion

```

File: docs/source/jit.md [hunk 22]
```diff
@@ -0,0 +1,901 @@
+# TorchScript
+
+```{eval-rst}
+.. toctree::
```
Comment (svekars) [inline docs/source/jit.md]: Please use eval-rst only for autodoc directives. MyST markdown has good [support](https://myst-parser.readthedocs.io/en/v0.15.1/syntax/syntax.html#syntax-directives-parsing) for toctree and most of other sphinx directives: 
```suggestion
```{toctree}
```

**Comment by thatgeeman** [comment]
@svekars I updated my fork from pytorch:main before making the changes suggested above. This PR Looks messy now. Shall I close this and make a new PR?
